transport/src/main/java/io/netty/channel/AbstractFileRegion.java

```diff
@@ -0,0 +1,184 @@
+        if (written > 0) {
+            transferIndex += written;
+        }
+        return written;
+    }
+
+    /**
+     * Returns the {@code transferIndex} of this file region.
+     */
+    public long transferIndex() {
+        return transferIndex;
+    }
+
+    /**
+     * Sets the {@code transferIndex} of file region.
+     *
+     * @throws IndexOutOfBoundsException
+     *             if the specified {@code transferIndex} is less than {@code 0} or greater than
+     *             {@link #count()}
+     */
+    public FileRegion transferIndex(long index) {
+        if (index < 0 || index > count()) {
+            throw new IndexOutOfBoundsException(
+                    String.format("transferIndex: %d (expected: 0 <= transferIndex <= count(%d))",
+                            index, count()));
+        }
+        this.transferIndex = index;
+        return this;
+    }
+
+    /**
+     * Returns {@code true} if and only if {@link #transferableBytes()} is greater than {@code 0}.
+     */
+    public boolean isTransferable() {
+        return transferIndex < count();
+    }
+
+    /**
+     * Returns the number of readable bytes which is equal to {@code (count - transferIndex)}.
+     */
+    public long transferableBytes() {
+        return count() - transferIndex;
+    }
+
+    /**
+     * Transfers the content of this file region to the specified channel.
+     * <p>
+     * This method does not modify {@link #transferIndex()}.
+     *
+     * @param target
+     *            the destination of the transfer
+     * @param position
+     *            the relative offset of the file where the transfer begins from. For example,
+     *            <tt>0</tt> will make the transfer start from {@link #position()}th byte and
+     *            <tt>{@link #count()} - 1</tt> will make the last byte of the region transferred.
+     * @param length
+     *            the maximum number of bytes to transfer
+     */
+    public long transferBytesTo(WritableByteChannel target, long position, long length)
+            throws IOException {
+        if (position < 0 || position > count()) {
+            throw new IllegalArgumentException("position out of range: " + position
+                    + " (expected: 0 - " + (this.count() - 1) + ')');
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("negative length " + length);
+        }
+        length = Math.min(count() - position, length);
+        if (length == 0) {
+            return 0L;
+        }
+        return channel().transferTo(position() + position, length, target);
+    }
+
+    /**
+     * Returns a slice of this file region's sub-region. This method does not modify
+     * {@code transferIndex} of this buffer.
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the reference count
+     * will NOT be increased.
+     */
+    public AbstractFileRegion slice(long index, long length) {
```

Comment:
Here we need the returned FileRegion also has the newly added method, otherwise we need to do instanceof and cast everywhere after a slice. We can change to `FileRegion` after moving the methods to FileRegion interface?